src/scalar_8x32_impl.h

```diff
@@ -718,4 +720,99 @@ SECP256K1_INLINE static void secp256k1_scalar_mul_shift_var(secp256k1_scalar *r,
     secp256k1_scalar_cadd_bit(r, 0, (l[(shift - 1) >> 5] >> ((shift - 1) & 0x1f)) & 1);
 }
 
+#define ROTL32(x,n) ((x) << (n) | (x) >> (32-(n)))
+#define QUARTERROUND(a,b,c,d) \
+  a += b; d = ROTL32(d ^ a, 16); \
+  c += d; b = ROTL32(b ^ c, 12); \
+  a += b; d = ROTL32(d ^ a, 8); \
+  c += d; b = ROTL32(b ^ c, 7);
+
+#ifdef WORDS_BIGENDIAN
+#define LE32(p) ((((p) & 0xFF) << 24) | (((p) & 0xFF00) << 8) | (((p) & 0xFF0000) >> 8) | (((p) & 0xFF000000) >> 24))
+#else
+#define LE32(p) (p)
+#endif
+
+static void secp256k1_scalar_chacha20(secp256k1_scalar *r1, secp256k1_scalar *r2, const unsigned char *seed, uint64_t idx) {
```

Comment:
A thought about multiple implementations of the same cryptography.
is this library becoming a general cryptography library?
should there be a different general crypto library with sha2/chacha20 etc.?

Because right now both libsecp and bitcoin implements SHA256 and HMAC, and after this they'll also both implement Chacha20 (https://github.com/bitcoin/bitcoin/blob/master/src/crypto/chacha20.cpp)

Any ideas on the matter? (I myself don't like duplicate implementations especially of sensitive cryptography(and there's also a performance difference too))

Reply:
libsecp256k1 is absolutely not a generic low-level cryptographc primitives library; it's a library implementing high-level cryptographic algorithms efficiently with a safe-to-use interface.

As far as duplicate implementations goes, these are included for necessity of exposing fully-functional versions of the high-level functions it implements. I personally don't think code duplication for such low-level primitives is an issue, as this code doesn't "rot" (it won't go out of date ever, at worst very rarely needs to be extended, and possibly doesn't need to be touched ever, at all), so it isn't subject to the same priorities most software engineering places.

That said, there are some practical reasons why having a built-in implementation may be undesirable:

* In very low-memory systems which already have versions of SHA256/... somewhere, the additional code size may be undesirable.
* The SHA256/Chacha20 code inside libsecp256k1 isn't very optimized, and probably never will gain high levels of architecture-specific optimization. Generally in these protocols, the performance of these is not very important, but if it is, it'd be great to have libsecp256k1 use more efficient versions.

Because of that reason, I think we should see the internal primitives as default, naive, private implementations, and if there is ever a reason to use another one, we should add functionality to make libsecp256k1 use external versions (rather than exposing/improving the internal ones).

Reply:
Yeah, I didn't consider the first option a real one, just a possibility about how will the library look with more primitives being added.

I guess your answer is fair, optimized sha256 might already be a nice speedup in a rapid verification scenario(which is already considered in the whole batch verification).

Maybe making a general low level library that is shared between libsecp and bitcoin is an option? Otherwise your suggestion that if someone really wanted then it's possible to add a feature to support these functions as extern.